static/app/utils/profiling/differentialFlamegraph.tsx

```diff
@@ -170,18 +136,109 @@ export class DifferentialFlamegraph extends Flamegraph {
     differentialFlamegraph.colorBuffer = makeColorBuffer(
       sourceFlamegraph.frames,
       colorMap,
-      theme.COLORS.FRAME_FALLBACK_COLOR as unknown as ColorChannels,
-      DifferentialFlamegraph.FrameKey
+      theme.COLORS.FRAME_FALLBACK_COLOR as unknown as ColorChannels
     );
 
     differentialFlamegraph.newFrames = newFrames;
     differentialFlamegraph.removedFrames = removedFrames;
     differentialFlamegraph.increasedFrames = increasedFrames;
     differentialFlamegraph.decreasedFrames = decreasedFrames;
-    differentialFlamegraph.beforeCounts = beforeCounts;
-    differentialFlamegraph.afterCounts = afterCounts;
+    differentialFlamegraph.weights = weights;
     differentialFlamegraph.negated = negated;
 
     return differentialFlamegraph;
   }
 }
+
+export function diffFlamegraphTreeRecursive(
+  beforeFlamegraph: Flamegraph,
+  afterFlamegraph: Flamegraph,
+  negated
+): {
+  newFrames: FlamegraphFrame[];
+  removedFrames: FlamegraphFrame[];
+  weights: Map<FlamegraphFrame['node'], {after: number; before: number}>;
+} {
+  let removedFrames: FlamegraphFrame[] = [];
+  let newFrames: FlamegraphFrame[] = [];
+
+  const weights = new Map<FlamegraphFrame['node'], {after: number; before: number}>();
+
+  function visit(beforeFrame: FlamegraphFrame, afterFrame: FlamegraphFrame) {
+    weights.set(negated ? beforeFrame.node : afterFrame.node, {
+      before: beforeFrame.node.totalWeight,
+      after: afterFrame.node.totalWeight,
+    });
+
+    const beforeFrameChildrenLength = beforeFrame.children.length;
+    const afterFrameChildrenLength = afterFrame.children.length;
+
+    // In case the current node has no children, we need to check if the
+    // other node has children, add them to the removedFrames list
+    if (!afterFrameChildrenLength && beforeFrameChildrenLength) {
+      for (let i = 0; i < beforeFrameChildrenLength; i++) {
+        removedFrames = removedFrames.concat(getTreeNodes(beforeFrame.children[i]));
+      }
+      return;
+    }
+
+    // In case the current node has children, but the other node doesn't,
+    // we mark the entire part of the tree as new
+    if (!beforeFrameChildrenLength && afterFrameChildrenLength) {
+      for (let i = 0; i < afterFrameChildrenLength; i++) {
+        newFrames = newFrames.concat(getTreeNodes(afterFrame.children[i]));
+      }
+      return;
+    }
+
+    for (let i = 0; i < afterFrameChildrenLength; i++) {
+      for (let j = 0; j < beforeFrameChildrenLength; j++) {
```

Comment:
Wouldn't this repeat a lot of added/removed frames since it loops over all before frames for all after frames?